### PR TITLE
Implement Roughtime client pseudo-node

### LINKED ISSUE #730: Support Roughtime from Rust
We currently have some Roughtime client code in C++ (#428, #461, #468).  It's not yet plumbed into the runtime (there's no ABI to expose it yet), but it will eventually need to be accessible from the Rust-based version of the runtime (#723).   That could either be:
 - because it's possible to link C++ functionality into the Rust runtime (#724), or
 - because we have a pure Rust implementation of the Roughtime client.

### PR DESCRIPTION
**Pull request by conradgrobler**
# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

Fixes #730

### CODE REVIEW

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 1]
```diff
@@ -16,32 +16,171 @@
...
+            ),
         }
     }
+
+    /// Processes a gRPC invocation.
+    fn process_invocation(&self, runtime: &RuntimeProxy, invocation: &Invocation) {
+        match invocation.receive_request(runtime) {
+            Ok(request) => {
+                if request.method_name != "/oak.roughtime.RoughtimeService/GetRoughtime" {
+                    let message = format!("Unknwon method_name: {}", request.method_name);
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/node/roughtime.rs:107]: s/Unknwon/Unknown/

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 2]
```diff
@@ -16,32 +16,171 @@
...
+
+        match self.client.get_roughtime() {
+            Ok(time) => {
+                let reply = RoughtimeResponse {
+                    roughtime_usec: time,
+                };
+                let mut message = Vec::new();
+                match reply.encode(&mut message) {
+                    Ok(_) => {
+                        let response = GrpcResponse {
```
Comment (daviddrysdale) [inline oak/server/rust/oak_runtime/src/node/roughtime.rs:132]: nit: having both `reply` and `response` is a bit confusing; maybe `rsp` and `grpc_rsp` or somesuch?

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 3]
```diff
@@ -16,32 +16,171 @@
...
+use crate::{
+    io::Receiver,
+    node::grpc::invocation::Invocation,
+    runtime::RuntimeProxy,
+    time::{
+        get_default_servers, RoughtimeClient, RoughtimeServer, DEFAULT_MAX_RADIUS_MICROSECONDS,
+        DEFAULT_MIN_OVERLAPPING_INTERVALS, DEFAULT_SERVER_RETRIES, DEFAULT_TIMEOUT_SECONDS,
+    },
+};
+
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak/server/rust/oak_runtime/src/node/roughtime.rs]: Removed.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 4]
```diff
@@ -16,32 +16,171 @@
...
+        google::rpc::Code,
+        oak::{
+            application::RoughtimeClientConfiguration,
+            encap::GrpcResponse,
+            roughtime::{RoughtimeRequest, RoughtimeResponse},
+        },
+    },
+    OakStatus,
+};
+use prost::Message as _;
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak/server/rust/oak_runtime/src/node/roughtime.rs]: Removed. I thought it was needed to stop cargo from complaining about unused imports, but cargo did not complain after removing it.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs (hunk 1 above)
Comment (conradgrobler) [reply oak/server/rust/oak_runtime/src/node/roughtime.rs]: Done.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs (hunk 2 above)
Comment (conradgrobler) [reply oak/server/rust/oak_runtime/src/node/roughtime.rs]: Renamed.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 5]
```diff
@@ -16,32 +16,170 @@
...
 
 impl RoughtimeClientNode {
     /// Creates a new [`RoughtimeClientNode`] instance, but does not start it.
-    pub fn new(node_name: &str) -> Self {
+    pub fn new(node_name: &str, config: &RoughtimeClientConfiguration) -> Self {
+        let timeout_seconds = if config.timeout_seconds == 0 {
+            DEFAULT_TIMEOUT_SECONDS
+        } else {
+            config.timeout_seconds as u64
+        };
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/roughtime.rs:56]: If these values are meant to be optional, then perhaps should use proper types in the protos, e.g. https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/u-int32-value, which prost should convert into `Option<u32>` , which you can then use as `config.server_retries.unwrap_or(DEFAULT_SERVER_RETRIES)` etc.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs [hunk 6]
```diff
@@ -16,32 +16,170 @@
...
+                server_retries,
+            ),
         }
     }
+
+    /// Processes a gRPC invocation.
+    fn process_invocation(&self, runtime: &RuntimeProxy, invocation: &Invocation) {
+        match invocation.receive_request(runtime) {
+            Ok(request) => {
+                if request.method_name != "/oak.roughtime.RoughtimeService/GetRoughtime" {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/roughtime.rs:105]: Should this be auto generated at some point? Do we have a GitHub issue to track that already?

File: oak/proto/roughtime_service.proto [hunk 7]
```diff
@@ -18,18 +18,18 @@ syntax = "proto3";
...
   // 86,399 SI seconds, and all the smeared seconds are the same length.
-  uint64 rough_time_usec = 1;
+  uint64 roughtime_usec = 1;
 }
 
 // Interface exposed by the Roughtime client pseudo-Node to other nodes over a
 // pair of Oak Channels.
 service RoughtimeService {
-  rpc GetRoughTime(RoughTimeRequest) returns (RoughTimeResponse);
+  rpc GetRoughtime(RoughtimeRequest) returns (RoughtimeResponse);
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply oak/proto/roughtime_service.proto]: Renamed.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs (hunk 5 above)
Comment (conradgrobler) [reply oak/server/rust/oak_runtime/src/node/roughtime.rs]: Done.

File: oak/server/rust/oak_runtime/src/node/roughtime.rs (hunk 6 above)
Comment (conradgrobler) [reply oak/server/rust/oak_runtime/src/node/roughtime.rs:98]: Created #1113

File: oak/proto/application.proto [hunk 8]
```diff
@@ -119,10 +121,10 @@ message RoughtimeClientConfiguration {
   repeated RoughtimeServer servers = 1;
   // Connection parameters; default values will be used if any parameter is
   // zero.
```
Comment (tiziano88) [inline oak/proto/application.proto:123]:
```suggestion
  // unset.
```

Comment (conradgrobler) [reply oak/proto/application.proto]: Done.

**Pull request merged**